xrspatial/zonal.py

```diff
@@ -657,14 +654,155 @@ def _crosstab_dask(
+        cat_ids = [c for c in cat_ids if c in unique_cats]
+    return unique_cats, cat_ids
+
+
+def _sort_values(sorted_indices, values, cats):
+    if len(values.shape) == 2:
+        # 2D case
+        result = values.flatten()[sorted_indices]
+    else:
+        # 3D case
+        num_cats = len(cats)
+        h, w = values.shape[1:]
+        result = np.zeros(shape=(num_cats, h * w), dtype=values.dtype)
+        for i, cat in enumerate(cats):
+            result[i] = values[i].flatten()[sorted_indices]
+
+    return result
+
+
+def _get_zone_values(values_by_zones, start, end):
+    if len(values_by_zones.shape) == 1:
+        # 1D flatten, i.e, original data is 2D
+        return values_by_zones[start:end]
+    else:
+        # 2D flatten, i.e, original data is 3D
+        return values_by_zones[:, start:end]
+
+
+def _single_zone_crosstab(
+    zone_values,
+    unique_cats,
+    cat_ids,
+    nodata_values,
+    crosstab_dict,
+    agg
+):
+    if len(zone_values.shape) == 1:
+        # 1D flatten, i.e, original data is 2D
+
+        # filter out non-finite and nodata_values
+        zone_values = zone_values[
+            np.isfinite(zone_values) & (zone_values != nodata_values)
+            ]
+        total_count = zone_values.shape[0]
+
+        sorted_zone_values = np.sort(zone_values)
+        zone_cat_breaks = _strides(sorted_zone_values, unique_cats)
+        cat_start = 0
+
+        for j, cat in enumerate(unique_cats):
+            if cat in cat_ids:
+                count = zone_cat_breaks[j] - cat_start
+                if agg == 'count':
+                    crosstab_dict[cat].append(count)
+                elif agg == 'percentage':
+                    pct = np.nan
+                    if total_count != 0:
+                        pct = count / total_count * 100
+                    crosstab_dict[cat].append(pct)
+                cat_start = zone_cat_breaks[j]
+
+    else:
+        # 2D flatten, i.e, original data is 3D
+        for j, cat in enumerate(unique_cats):
+            if cat in cat_ids:
+                zone_cat_data = zone_values[j]
+                # filter out non-finite and nodata_values
+                zone_cat_data = zone_cat_data[
+                    np.isfinite(zone_cat_data)
+                    & (zone_cat_data != nodata_values)
+                ]
+                count = zone_cat_data.shape[0]
+                crosstab_dict[cat].append(count)
+
+
+def _faster_crosstab_numpy(
+    zones: np.ndarray,
+    values: np.ndarray,
+    zone_ids: List[Union[int, float]],
+    unique_cats: np.ndarray,
+    cat_ids: Union[List, np.ndarray],
+    nodata_values: Union[int, float],
+    agg: str,
+) -> pd.DataFrame:
+    # find ids for all zones
+    unique_zones = np.unique(zones[np.isfinite(zones)])
+    # selected zones to do analysis
+    if zone_ids is None:
+        zone_ids = unique_zones
+    # remove zones that does not exist in `zones` raster
+    zone_ids = [z for z in zone_ids if z in unique_zones]
```

Comment:
Could this be in an `else` block of the previous `if` statement? (Like first review comment).